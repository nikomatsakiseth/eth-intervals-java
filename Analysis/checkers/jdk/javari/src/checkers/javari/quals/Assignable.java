package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that a field is assignable, even if it is inside a {@link ReadOnly}
 * instance.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target({FIELD})
public @interface Assignable {

}
