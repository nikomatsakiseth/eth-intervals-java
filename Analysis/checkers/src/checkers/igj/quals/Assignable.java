package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import static java.lang.annotation.ElementType.*;

/**
 *
 * Indicates the annotated {@code Field} may be re-assigned regardless of the
 * immutability of the enclosing class or object instance.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
@Target( { FIELD } )
public @interface Assignable {

}
