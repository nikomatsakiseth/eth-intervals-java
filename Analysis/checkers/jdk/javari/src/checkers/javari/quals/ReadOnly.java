package checkers.javari.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.javari.JavariChecker;
import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import static java.lang.annotation.ElementType.*;

/**
 * Indicates that, for the variable on which this annotation appears,
 * the object to which this variable refers will not be modified via
 * this reference, except its fields explicitly marked as
 * {@link Mutable}.
 *
 * <p>
 *
 * This annotation is part of the Javari language.
 *
 * @see Mutable
 * @see JavariChecker
 * @checker.framework.manual #javari-checker Javari Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target({FIELD, LOCAL_VARIABLE, METHOD, PARAMETER, TYPE})
@TypeQualifier
@SubtypeOf({})
public @interface ReadOnly {

}
