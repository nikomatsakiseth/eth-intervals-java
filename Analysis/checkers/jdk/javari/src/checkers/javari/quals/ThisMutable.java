package checkers.javari.quals;

import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 * An annotation used to represent a place holder immutability type, that is
 * equivalent to the ThisMutable type on the Javari typesystem.
 *
 * However, it is an implementation detail; hence, the package-scope.
 */
@TypeQualifier
@Target({}) // A programmer cannot write @ThisMutable in a program
@SubtypeOf(ReadOnly.class)
public @interface ThisMutable {

}
