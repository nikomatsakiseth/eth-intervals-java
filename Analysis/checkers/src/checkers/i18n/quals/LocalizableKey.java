package checkers.i18n.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;
import checkers.quals.Unqualified;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface LocalizableKey { }
