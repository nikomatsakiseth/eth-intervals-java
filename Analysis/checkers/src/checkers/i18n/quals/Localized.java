package checkers.i18n.quals;

import static com.sun.source.tree.Tree.Kind.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

import checkers.quals.*;

/**
 * Indicates that the {@code String} type has been localized and
 * formatted for the target output locale
 */
@TypeQualifier
@SubtypeOf(Unqualified.class)
@ImplicitFor( trees = {
        /* All integer literals */
        INT_LITERAL,
        LONG_LITERAL,
        FLOAT_LITERAL,
        DOUBLE_LITERAL,
        BOOLEAN_LITERAL,

        //CHAR_LITERAL,
        //STRING_LITERAL,
        //NULL_LITERAL
        }
)
@Documented
@Retention(RetentionPolicy.RUNTIME)
public @interface Localized { }
