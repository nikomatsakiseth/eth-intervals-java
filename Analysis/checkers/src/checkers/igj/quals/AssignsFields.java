package checkers.igj.quals;

import static java.lang.annotation.ElementType.*;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.SubtypeOf;
import checkers.quals.TypeQualifier;

/**
 *
 * Indicates that the annotated method could assigns (but not mutate) the fields
 * of {@code this} object.
 *
 * A method with an AssignsFields receiver may not use the receiver to
 * invoke other methods with mutable or immutable reciever.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
// TODO: Document this

@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { CONSTRUCTOR, METHOD })
@TypeQualifier // (for now)
@SubtypeOf( ReadOnly.class )
public @interface AssignsFields {

}
