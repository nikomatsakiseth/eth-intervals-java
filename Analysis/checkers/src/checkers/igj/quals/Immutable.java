package checkers.igj.quals;

import java.lang.annotation.Documented;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

import checkers.quals.*;

import static java.lang.annotation.ElementType.*;

/**
 * Indicates that the annotated reference is an immutable reference to an
 * immutable object.
 *
 * An Immutable object cannot be modified. Its fields may be reassigned or
 * mutated only if they are explicitly marked as Mutable or Assignable.
 *
 * @checker.framework.manual #igj-checker IGJ Checker
 */
@Documented
@Retention(RetentionPolicy.RUNTIME)
//@Target( { FIELD, LOCAL_VARIABLE, CONSTRUCTOR, METHOD, PARAMETER, TYPE })
@TypeQualifier
@SubtypeOf( ReadOnly.class )
public @interface Immutable {

}
