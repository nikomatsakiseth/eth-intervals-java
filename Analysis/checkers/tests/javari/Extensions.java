import checkers.javari.quals.*;

class Extensions {

    class BaseClass {
        String mutableReceiver() {return null;}
        String readonlyReceiver() @ReadOnly {return null;}
        @Mutable String mutableReturn() {return null;}
        @PolyRead String polyReadReturn() @PolyRead {return null;}
        @ReadOnly String readonlyReturn() {return null;}
    }

    @ReadOnly class ReadOnlyEmpty {}

    @ReadOnly class ReadOnlyExtension extends BaseClass {}

    @Mutable class MutableExtension extends BaseClass {}

    @PolyRead class PolyReadIllegalExtension extends BaseClass {}

    class SwitchReceivers extends BaseClass {
        String mutableReceiver() @ReadOnly {return null;}
        String readonlyReceiver() @Mutable {return null;} // error
    }

    class SwitchReturns extends BaseClass {
        @ReadOnly String mutableReturn() {return null;} // error
        @Mutable String readonlyReturn() {return null;}
    }

    class PolyReadReturns extends BaseClass {
        @PolyRead String mutableReturn() {return null;} // error
        @PolyRead String readonlyReturn() {return null;}
    }

    class ForcePolyReadReturnToMutable extends BaseClass {
        @Mutable String polyReadReturn() @ReadOnly {return null;}
    }

    class ForcePolyReadReturnToReadOnly extends BaseClass {
        @ReadOnly String polyReadReturn() @ReadOnly {return null;}
    }

    class PolyReadReceivers extends BaseClass {
        String mutableReceiver() @PolyRead {return null;}
        String readonlyReceiver() @PolyRead {return null;} // error
    }

}
