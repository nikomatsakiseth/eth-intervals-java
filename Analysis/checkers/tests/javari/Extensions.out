:19: (polyread.type)
:23: (override.receiver.invalid)
:27: (override.return.invalid)
:32: (override.return.invalid)
:41: (override.return.invalid)
:46: (override.receiver.invalid)
